### ISSUE TITLE
[HOLD for payment 2024-06-20] [$500] [Held requests] No confirmation prompt when approving held request via the report preview

### ISSUE THREAD

**Issue by lanitochka17** [Expensify/App#39335]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/contributingGuides/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

**Version Number:** 1.4.58-4
**Reproducible in staging?:** Y
**Reproducible in production?:** N
**If this was caught during regression testing, add the test name, ID and link from TestRail:** N/A
**Issue reported by:** Applause - Internal Team

**Issue found when executing PR** https://github.com/Expensify/App/pull/33124
## Action Performed:
Precondition:
- Admin and employee are in the Collect workspace
1. Go to staging.new.expensify.com
2. [Employee] Request money from the workspace chat
3. [Employee] Go to transaction thread, select Hold request and hold the request
4. [Admin] Go to expense report with the employee
5. [Admin] Click Approve button
6. Note that it shows the prompt to approve the held request
7. [Admin] Return to the main workspace chat with employee
8. [Admin] Click Approve
9. Note that there is no prompt to approve the held request
10. Repeat Step 4 to 9 with Pay with Expensify button in both expense report and main chat

## Expected Result:
When clicking on the approve button in expense report (Step 4) and main chat (Step 7), there will be a confirmation prompt to approve the held request

## Actual Result:
When approving the held request from the main chat, there is no confirmation prompt. The same goes for Pay with Expensify button

## Workaround:
Unknown

## Platforms:

Which of our officially supported platforms is this issue occurring on?
- [x] Android: Native
- [x] Android: mWeb Chrome
- [x] iOS: Native
- [x] iOS: mWeb Safari
- [x] MacOS: Chrome / Safari
- [x] MacOS: Desktop

## Screenshots/Videos

Add any screenshot/video evidence

</details> 

https://github.com/Expensify/App/assets/78819774/d9443121-3cdb-4248-a9bb-f4fc16763cca

[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)

<details><summary>Upwork Automation - Do Not Edit</summary>
    <ul>
        <li>Upwork Job URL: https://www.upwork.com/jobs/~01db46b9f151c17d73</li>
        <li>Upwork Job ID: 1774864170413936640</li>
        <li>Last Price Increase: 2024-05-28</li>
<li>Automatic offers: </li>
<ul>
<li>DylanDylann | Contributor | 102499792</li>
<li>nkdengineer | Contributor | 102527527</li>
</ul></li>
    </ul>
</details>

<details><summary>Issue Owner</summary>Current Issue Owner: @slafortune</details>

**Comment by lanitochka17** [comment]
@bondydaa  FYI I haven't added the External label as I wasn't 100% sure about this issue. Please take a look and add the label if you agree it's a bug and can be handled by external contributors

**Comment by lanitochka17** [comment]
We think that this bug might be related to #wave-collect - Release 1

**Comment by bondydaa** [comment]
digging into this now

**Comment by bondydaa** [comment]
I think it coming from this PR https://github.com/Expensify/App/pull/33124/ but it's not immediately clear to me why, looking further.

**Comment by bondydaa** [comment]
oh i missed the exact case here, it's that clicking approve in the reportPreview doesn't trigger the model. I thought it was that when you click the same approve button twice it didn't re-trigger the confirmation model 🤦 .

**Comment by bondydaa** [comment]
I've asked for confirmation here https://github.com/Expensify/App/issues/31345#issuecomment-2030144285

not sure this actually needs to be a blocker or not, this case might not have been scoped out in the original GH so it could have just been an oversight.

**Comment by robertjchen** [comment]
Not a blocker, but definitely should be addressed!

**Comment by nkdengineer** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.
When approving the held request from the main chat, there is no confirmation prompt. The same goes for Pay with Expensify button

### What is the root cause of that problem?
In `ReportPreview`, we don't pass `confirmApproval` prop into `SettlementButton` so the default behavior is approve the request without confirmation prompt

https://github.com/Expensify/App/blob/99cc0b6f34c477dee40934b0ffb294e5dde1f58a/src/components/ReportActionItem/ReportPreview.tsx#L309

https://github.com/Expensify/App/blob/99cc0b6f34c477dee40934b0ffb294e5dde1f58a/src/components/SettlementButton.tsx#L218-L225
### What changes do you think we should make in order to solve the problem?
We can re-use the same approach in `MoneyReportHeader` to implement for both case payment and approve by using `ProcessMoneyReportHoldMenu` in `ReportPreivew`

1. In `ReportPreview`, create some state to store paymentType, visible hold menu, request type 

```
const [isHoldMenuVisible, setIsHoldMenuVisible] = useState(false);
const [requestType, setRequestType] = useState<'pay' | 'approve'>();
const [paymentType, setPaymentType] = useState<PaymentMethodType>();
const {isSmallScreenWidth} = useWindowDimensions();
```

2. Create `confirmPayment` and `confirmApproval` funtion

```
const confirmPayment = (type?: PaymentMethodType | undefined) => {
    if (!type) {
        return;
    }
    setPaymentType(type);
    setRequestType('pay');
    if (ReportUtils.hasHeldExpenses(iouReport?.reportID)) {
        setIsHoldMenuVisible(true);
    } else if (chatReport) {
        IOU.payMoneyRequest(type, chatReport, iouReport as Report, false);
    }
};

const confirmApproval = () => {
    setRequestType('approve');
    if (ReportUtils.hasHeldExpenses(iouReport?.reportID)) {
        setIsHoldMenuVisible(true);
    } else {
        IOU.approveMoneyRequest(iouReport as Report, true);
    }
};
```

3. Pass `confirmPayment` function as `onPress` prop of  `SettlementButton` and pass `confirmApproval` prop into it

```
onPress={confirmPayment}
confirmApproval={confirmApproval}
```
https://github.com/Expensify/App/blob/99cc0b6f34c477dee40934b0ffb294e5dde1f58a/src/components/ReportActionItem/ReportPreview.tsx#L314

4. Add `ProcessMoneyReportHoldMenu` component

```
{isHoldMenuVisible && requestType !== undefined && (
    <ProcessMoneyReportHoldMenu
        nonHeldAmount={!ReportUtils.hasOnlyHeldExpenses(iouReport?.reportID ?? '') ? nonHeldAmount : undefined}
        requestType={requestType}
        fullAmount={fullAmount}
        isSmallScreenWidth={isSmallScreenWidth}
        onClose={() => setIsHoldMenuVisible(false)}
        isVisible={isHoldMenuVisible}
        paymentType={paymentType}
        chatReport={chatReport}
        moneyRequestReport={iouReport as Report}
    />
)}
```
### What alternative solutions did you explore? (Optional)
NA

**Comment by slafortune** [comment]
@abdulrahuman5196 will you get a chance to review this in the next day or two?

**Comment by abdulrahuman5196** [comment]
Hi, Sorry, I was under the impression this issue was on hold from the title. My bad 😶‍🌫️ .

Will work on action items within a day.

**Comment by mvtglobally** [comment]
Issue not reproducible during KI retests. (First week)

**Comment by abdulrahuman5196** [comment]
Checking now

**Comment by bondydaa** [comment]
if we can't reproduce anymore we can probably just close out, right?

**Comment by nkdengineer** [comment]
I still can reproduce this one.

**Comment by abdulrahuman5196** [comment]
@mvtglobally Could you kindly share video if the issue is not reproducible?

**Comment by slafortune** [comment]
@abdulrahuman5196 if you are able to reproduce this, feel free to share a video or we can close it this.

**Comment by slafortune** [comment]
I will be out until 4/22 but not assigning a new BZ since this seems to not be reproducible - we'll give it one more week.

If someone disagrees, please add a video following the OG steps.

**Comment by nkdengineer** [comment]
@abdulrahuman5196 @slafortune I'm still able to reproduce this bug on staging or latest main.


https://github.com/Expensify/App/assets/161821005/b789f814-b546-47a4-9fa9-fde5db8ece97

**Comment by bondydaa** [comment]
@abdulrahuman5196 sounds like others can still reproduce so let's get proposals reviewed then

**Comment by abdulrahuman5196** [comment]
@bondydaa / @slafortune How do i enable this Hold Request feature?  For me its not present in main?

<img width="1512" alt="Screenshot 2024-04-22 at 9 18 55 PM" src="https://github.com/Expensify/App/assets/46707890/c9ea17e4-a7b5-4c2b-aa39-5a78887bea55">

**Comment by slafortune** [comment]
@bondydaa do we put much weight on reproducing on main?

@abdulrahuman5196 are you able to reproduce this on staging?

**Comment by nkdengineer** [comment]
@abdulrahuman5196 As the video you're in combine report which will be displayed if we have only one transaction in IOU report. You can create another one in this report and then you can go to transaction thread report to hold the request.

**Comment by slafortune** [comment]
Thanks @nkdengineer  - @abdulrahuman5196 Are you able to repo this on staging?  I'd like to get this moving forward or close if needed.

**Comment by bondydaa** [comment]
thanks @nkdengineer for the clarification, so you need a report with at least 2 transactions attached in order to be able to "hold" the report then?

**Comment by nkdengineer** [comment]
@bondydaa Yes.

**Comment by bondydaa** [comment]
@mvtglobally @abdulrahuman5196 can you retest again ensuring the report has at least 2+ transactions/expenses on it so that it's able to be held and lets confirm that the bug still exists?

**Comment by abdulrahuman5196** [comment]
I was able to put the transaction on hold. But after that there is no `Approve` flow, I am only getting `Submit`.

@mvtglobally / @nkdengineer  Can you check if the bug still exists or you are observing the same


https://github.com/Expensify/App/assets/46707890/76a59912-fd49-491f-837e-23cbd5858190

**Comment by nkdengineer** [comment]
@abdulrahuman5196 You need to submit first and then we can approve the request.

**Comment by abdulrahuman5196** [comment]
> @abdulrahuman5196 You need to submit first and then we can approve the request.

I tried that as well. After submit also I didn't get Approve step.

**Comment by slafortune** [comment]
@abdulrahuman5196 do you have an update?

**Comment by abdulrahuman5196** [comment]
> @abdulrahuman5196 do you have an update?

@slafortune  I am having hard time reproducing this issue. Can you try from your end?

**Comment by bondydaa** [comment]
@mvtglobally can you also retest for us? note these points https://github.com/Expensify/App/issues/39335#issuecomment-2083210732 / https://github.com/Expensify/App/issues/39335#issuecomment-2098381504

**Comment by bondydaa** [comment]
pinged in the QA channel to get this rested https://expensify.slack.com/archives/C9YU7BX5M/p1715809626074589

**Comment by mvtglobally** [comment]
@bondydaa We are getting mixed results on this one. some team members were not able to repro and some did repro 

https://github.com/Expensify/App/assets/43995119/fb7dadcd-4fbc-49ce-bac6-fab88287d759


https://github.com/Expensify/App/assets/43995119/51248d13-64e3-40ca-9a0a-a08a78b4e781

**Comment by slafortune** [comment]
@bondydaa do you think we should close this as it's not able to consistently be reproduced?

**Comment by nkdengineer** [comment]
@mvtglobally The bug here is no confirmation prompt when we approve held request via the report preview not expense report and I see in the video above it's reproducible.

**Comment by bondydaa** [comment]
@abdulrahuman5196 do you agree with @nkdengineer 's proposal or have any issues with it? https://github.com/Expensify/App/issues/39335#issuecomment-2030358362

I'm not sure we should close it and if https://github.com/Expensify/App/issues/39335#issuecomment-2030358362 is able to fix it consistently then lets do it.

**Comment by JmillsExpensify** [comment]
I think there might be some confusion. This happens consistently in the workspace chat report preview. It wasn't part of the original design, but it was a feature request from @puneetlath

**Comment by trjExpensify** [comment]
Yeah, this is a feature request. More succinct steps might be: 

1. Submit two requests to a workspace
2. As the approver, click into the expense report and put one on hold
3. Back out to the parent workspace chat
4. Click "Approve" on the expense report preview component 
5. Observe there's no confirmation modal

**Comment by bondydaa** [comment]
@JmillsExpensify @trjExpensify does this being a feature request then change anything with how we should handle it? 

I guess we should remove the `bug` label? Anything else?

**Comment by trjExpensify** [comment]
Nope, we can switch to the `New feature` label but it makes no material difference on the next steps. 

1. @abdulrahuman5196 - please review this [proposal](https://github.com/Expensify/App/issues/39335#issuecomment-2030358362)
2. After which, @bondydaa you can agree with that or not 
3. All being well, @nkdengineer can create a PR for it

**Comment by slafortune** [comment]
I will be out on vacation until June 4th - so adding another BZ in case there is a need in the meantime.  I'll take it back on the 4th.

**Comment by abdulrahuman5196** [comment]
Checking now

**Comment by abdulrahuman5196** [comment]
@strepanier03 / @bondydaa I am still having hard time reproducing this issue. I am fine with changing C+ who can reproduce this issue and take over for the issue to move faster. Kindly unassign me and add another C+.

**Comment by DylanDylann** [comment]
I can take over this issue as C+

**Comment by trjExpensify** [comment]
Done! @DylanDylann can you review the proposal today, please? Thanks!

**Comment by DylanDylann** [comment]
I will complete my review on May 29th

**Comment by DylanDylann** [comment]
I still can reproduce on latest main

**Comment by DylanDylann** [comment]
@GandalfGwaihir @nkdengineer Both guys mentioned the correct root cause. 

@GandalfGwaihir Besides passing `confirmApproval` to `SettlementButton` component we also need to add `ProcessMoneyReportHoldMenu` modal to `ReportPreview` as included in @nkdengineer's proposal

@nkdengineer's [proposal](https://github.com/Expensify/App/issues/39335#issuecomment-2030358362) looks good to me

🎀👀🎀 C+ reviewed

**Comment by trjExpensify** [comment]
Perfect. @bondydaa, you good with the proposal? Can we move on to a PR? Thanks!

**Comment by nkdengineer** [comment]
@DylanDylann The PR is [here](https://github.com/Expensify/App/pull/42896).

**Comment by slafortune** [comment]
Thanks @strepanier03 - I'll take this back now, vacation is O V E R!

**Comment by DylanDylann** [comment]
BugZero Checklist: The PR fixing this issue has been merged! The following checklist ([instructions](https://stackoverflow.com/c/expensify/questions/14574)) will need to be completed before the issue can be closed:

 [@DylanDylann] The PR that introduced the bug has been identified. Link to the PR: **NA**
 [@DylanDylann] The offending PR has been commented on, pointing out the bug it caused and why, so the author and reviewers can learn from the mistake. Link to comment: **NA**
 [@DylanDylann] A discussion in #expensify-bugs has been started about whether any other steps should be taken (e.g. updating the PR review checklist) in order to catch this type of bug sooner. Link to discussion: **NA**
 [@DylanDylann] Determine if we should create a regression test for this bug. **Yes**
 [@DylanDylann] If we decide to create a [regression test](https://github.com/Expensify/App/blob/main/contributingGuides/REGRESSION_TEST_BEST_PRACTICES.md) for the bug, please propose the regression test steps to ensure the same bug will not reach production again.


### Regression Test Proposal

**Precondition: Workspace enabled approval**
1. Go to the WS chat
2. Create two expenses
3. Open the transaction thread report of one expense and hold the request
4. Submit the expense
5. As an approver, open WS chat and click on approve in the report preview
6. Verify that confirmation prompt appears


Do we agree 👍 or 👎

**Comment by slafortune** [comment]
@DylanDylann Paid via Upworks ✅ 
@nkdengineer Paid via Upworks ✅
